File: Episode12-Leaderboard/Monogame/Shield.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using MonoGame.Extended;

namespace Shmup
{
    internal static class Shield
    {
        /// <summary>
        /// draw 3 rectangles:
	    /// 1. solid background colour -> shared.DARK_BLUE
	    /// 2. solid green(width set by shield value)
	    /// 3. outline in white
        /// </summary>
        private static float percent = 100f;
        private static int barLength = 100;
        private static int barHeight = 10;
        private static Color lineColor = Color.White;
        private static Color fillColor = new Color(0, 0, 12);
        private static RectangleF outlineRect = new RectangleF(5, 5, barLength, barHeight);
        private static RectangleF fillRect = new RectangleF(5, 5, barLength, barHeight);
        private static RectangleF barRect = new RectangleF(5, 5, barLength, barHeight);

        public static void Update(float value)
        {
            if (value < 0) value = 0;
            percent = value;
            fillRect.Width = (percent / 100) * barLength;
        }
        public static void Draw(SpriteBatch spriteBatch)
        {
            spriteBatch.FillRectangle(barRect, fillColor);
            spriteBatch.FillRectangle(fillRect, Shared.GREEN);
            spriteBatch.DrawRectangle(outlineRect, lineColor, 2); // 2 pixel thick outline
        }
    }
}

File: Episode12-Leaderboard/Pygame/shield.py
import pygame, shared

class Shield():
	'''
	draw 3 rectangles:
	1. solid background colour -> shared.DARK_BLUE
	2. solid green (width set by shield value)
	3. outline in white
	'''
	def __init__(self, x:int, y:int, line_colour:tuple, fill_colour:tuple) -> None:
		self.BAR_LENGTH:int      = 100
		self.BAR_HEIGHT:int      = 10
		self.line_colour:tuple   = line_colour
		self.fill_colour:tuple   = fill_colour
		self.back_colour:tuple	 = shared.DARK_BLUE
		self.pct:int             = 0
		# create 3 rectangles
		self.bg_rect:object      = pygame.Rect(x, y, self.BAR_LENGTH, self.BAR_HEIGHT)
		self.outline_rect:object = self.bg_rect.copy()	# copy the rectangle else it refers to same rectangle!
		self.fill_rect:object    = self.bg_rect.copy()
		
	def update(self, value) -> None:
		''' change the bar length as a proportion of player.shield '''
		if value < 0:
			value = 0	
		self.pct = value
		self.fill_rect.width = (self.pct / 100) * self.BAR_LENGTH
	
	def draw(self) -> None:
		pygame.draw.rect(shared.screen, self.back_colour, self.bg_rect)			# dark background
		pygame.draw.rect(shared.screen, self.fill_colour, self.fill_rect)		# green shield value
		pygame.draw.rect(shared.screen, self.line_colour, self.outline_rect, 2) # 2 pixel thick outline